dm/syncer/validate_worker.go

```diff
@@ -66,118 +70,179 @@ type validateWorker struct {
 	change := vw.pendingChangesMap[fullTableName]
 	if change == nil {
 		// no change of this table
-		change = &tableChange{
-			table: row.table,
-			rows:  make(map[string]*rowChange),
-		}
+		change = newTableChange(row.table)
 		vw.pendingChangesMap[fullTableName] = change
 	}
-	if val, ok := change.rows[row.key]; ok {
-		val.data = row.data
-		val.tp = row.tp
-		val.lastMeetTS = row.lastMeetTS
-		val.failedCnt = 0 // clear failed count
+	if val, ok := change.rows[row.Key]; ok {
+		val.Data = row.Data
+		val.Tp = row.Tp
+		val.FirstValidateTS = 0
+		val.FailedCnt = 0 // clear failed count
 	} else {
-		change.rows[row.key] = row
+		change.rows[row.Key] = row
 		vw.pendingRowCount.Inc()
 	}
 }
 
-func (vw *validateWorker) validateTableChange() error {
-	failedChanges := make(map[string]*tableChange)
-	var failedRowCnt int64
+func (vw *validateWorker) validateTableChange() {
+	var err error
+	defer func() {
+		if err != nil {
+			// todo: better error handling
+			vw.validator.errChan <- terror.Annotate(err, "failed to validate table change")
+		}
+	}()
+
+	// clear accumulated row counter
+	vw.accuRowFromChanel.Store(0)
+
+	failedChanges := make(map[string]map[string]*validateFailedRow)
 	for k, tblChange := range vw.pendingChangesMap {
 		var insertUpdateChanges, deleteChanges []*rowChange
 		for _, r := range tblChange.rows {
-			if r.tp == rowDeleted {
+			if r.Tp == rowDeleted {
 				deleteChanges = append(deleteChanges, r)
 			} else {
 				insertUpdateChanges = append(insertUpdateChanges, r)
 			}
 		}
-		rows := make(map[string]*rowChange)
-		if len(insertUpdateChanges) > 0 {
-			failedRows, err := vw.validateRowChanges(tblChange.table, insertUpdateChanges, false)
-			if err != nil {
-				return err
+		allFailedRows := make(map[string]*validateFailedRow)
+		validateFunc := func(rows []*rowChange, isDelete bool) error {
+			if len(rows) == 0 {
+				return nil
 			}
-			// todo: if row failed count > threshold, should mark it as data inconsistent error
-			for pk := range failedRows {
-				rows[pk] = tblChange.rows[pk]
-				rows[pk].failedCnt++
+			failedRows, err2 := vw.validateRowChanges(tblChange.table, rows, isDelete)
+			if err2 != nil {
+				return err2
 			}
-		}
-		if len(deleteChanges) > 0 {
-			failedRows, err := vw.validateRowChanges(tblChange.table, deleteChanges, true)
-			if err != nil {
-				return err
+			for key, val := range failedRows {
+				allFailedRows[key] = val
 			}
-			for pk := range failedRows {
-				rows[pk] = tblChange.rows[pk]
-				rows[pk].failedCnt++
+			return nil
+		}
+		if err = validateFunc(insertUpdateChanges, false); err != nil {
+			return
+		}
+		if err = validateFunc(deleteChanges, true); err != nil {
+			return
+		}
+		if len(allFailedRows) > 0 {
+			failedChanges[k] = allFailedRows
+		}
+	}
+
+	vw.updatePendingAndErrorRows(failedChanges)
+}
+
+func (vw *validateWorker) updatePendingAndErrorRows(failedChanges map[string]map[string]*validateFailedRow) {
+	vw.Lock()
```

Comment:
I'm a bit curious about what this lock used for? It seems all the worker functions are executed synchronously in one go routine? Will they called by other outside code (e.g. DataValidator)?

Reply:
yes, validator routine will access it's fields